lua/AI/AIBuilders/AIExperimentalBuilders.lua

```
@@ -217,6 +220,7 @@ BuilderGroup {
             { UCBC, 'HaveLessThanUnitsInCategoryBeingBuilt', { 1, categories.EXPERIMENTAL}},
             { UCBC, 'HaveLessThanUnitsInCategoryBeingBuilt', { 1, categories.NUKE * categories.STRUCTURE}},
             { UCBC, 'HaveGreaterThanUnitsWithCategory', { 0, categories.ENERGYPRODUCTION * categories.TECH3}},
+            { EBC, 'GreaterThanMassStorageCurrent', { 150 }},
```

If I am correct the conditions are evaluated one-by-one, where the builder is rejected the moment it finds an invalid condition. From a performance perspective it is better to fail on simple checks quick, before running more expensive checks.

Without asking you to re-order all the conditions, could you add in future conditions with respect to how expensive they are? The average economy condition is (very) cheap in comparison to unit conditions: they should be at the top of the list.

---

Yes your right, line 223 is cheaper than any of the conditions on 220,221,222. I did reorder a bunch of the others but I also missed some. I'll try and reorder what I've got here.
The general rule is to have a happy medium between making sure the most likely to fail condition will come first, but also trying to have the cheapest first.

In this case GreaterThanMassStorageCurrent is one function call vs many for the HaveLessThanUnitsInCategoryBeingBuilt condition.

---

I think the order should depend on the complexity of the condition. As an example:

- `{ EBC, 'GreaterThanMassStorageCurrent', { 150 }},` has a complexity of O(1), because it doesn't rely on any list or other recursive datastructure
- `{ UCBC, 'HaveLessThanUnitsInCategoryBeingBuilt', { 1, categories.NUKE * categories.STRUCTURE}},` has a complexity of O(n) because internally it relies on `aiBrain:GetListOfUnits(categories.CONSTRUCTION, false)`, and then iterates over the results. A list (or table) is a recursive structure, meaning it can grow and shrink: these conditions get more and more expensive as the game progresses, as `n` relies on the number of units.

All unit conditions should be at the bottom (checked last), even if they would usually fail fast too (it still has to iterate over all the current units to determine that no unit matches a category!), while checking one value is always cheap.

---

i've reordered all conditions in the ones that I've changed. Thought it was better to just do it right the first time.